File: ingest_exercise.go
package main

import (
	"database/sql"
	"encoding/csv"
	"flag"
	"io"
	"log"
	"os"
	"path/filepath"

	_ "github.com/mattn/go-sqlite3"
)

func main() {
	fileToIngest := flag.String("file", "someFile.csv", "string: relative path to file")
	dbName := flag.String("dbName", "myDatabase.db", "string: relative path to sqlite db file")

	flag.Parse()

	// extract
	path, _ := filepath.Abs(*fileToIngest)
	f, err := os.Open(path)
	if err != nil {
		log.Fatal(err)
	}

	defer f.Close()

	records := csvToMap(f)
	_ = records

	// transform
	// ... your code here

	// load
	db, err := sql.Open("sqlite3", *dbName)
	if err != nil {
		log.Fatal(err)
	}
	defer db.Close()

	// ... insert the transformed records into your sqlite schema
}

// csvToMap takes a reader and returns an array of dictionaries, using the header row as the keys
func csvToMap(reader io.Reader) []map[string]interface{} {
	r := csv.NewReader(reader)
	var rows []map[string]interface{}
	header := make([]string, 0)
	for {
		record, err := r.Read()
		if err == io.EOF {
			break
		}
		if err != nil {
			log.Fatal(err)
		}
		if len(header) == 0 {
			header = record
		} else {
			dict := make(map[string]interface{})
			for i := range header {
				dict[header[i]] = record[i]
			}
			rows = append(rows, dict)
		}
	}
	return rows
}

File: ingest_exercise.py
import argparse
import sqlite3
from csv import DictReader


def main():
    parser = argparse.ArgumentParser()
    parser.add_argument('--ingest-filename', dest='ingest_filename', help='the relative path to the data file we are ingesting')
    parser.add_argument('--db-filename', dest='db_filename', help='the relative path to the database file we are inserting data in')
    args = parser.parse_args()

    # extract
    records = csv_to_map(args.ingest_filename)

    # transform
    # ... your code here

    # load
    try:
        sqlite_conn = sqlite3.connect(args.db_filename)
        cursor = sqlite_conn.cursor()
        cursor.close()
    except sqlite3.Error as error:
        print("Error loading table", error)
    finally:
        if sqlite_conn:
            sqlite_conn.close()

    # ... insert the transformed records into your sqlite schema


# csv_to_map takes a relative path to a csv file and returns a list of dictionaries, using the header row as the keys
def csv_to_map(ingest_filename):
    records = []
    with open(ingest_filename, 'r') as ingest_file:
        csv_dict_reader = DictReader(ingest_file)
        for row in csv_dict_reader:
            records.append(row)
    return records


if __name__ == "__main__":
    main()
